### PR TITLE
#22 - Support for root resource routing

### PR DESCRIPTION
**Pull request by olegmoz**
Part of issue #22 
Added routing to root path to support pushing a package to this path by PUT method.

### CODE REVIEW

**Comment by paulodamaso** [comment]
@g4s8 Please assign me

**Comment by g4s8** [comment]
@0crat assign @paulodamaso

**Comment by 0crat** [comment]
> @0crat assign @paulodamaso  ([here](https://github.com/artipie/composer-adapter/issues/32#issuecomment-593452032))

@g4s8 This pull request [#32](https://github.com/artipie/composer-adapter/issues/32) is assigned to @paulodamaso[/z](https://www.0crat.com/u/paulodamaso), here is [why](https://www.0crat.com/footprint/CT2E6TK9B/6d19c432-aafc-4f03-9f95-98c24834cf26); the budget is 15 minutes, see [§4](http://www.zerocracy.com/policy.html#4); please, read [§27](http://www.zerocracy.com/policy.html#27) and when you decide to accept the changes, inform @g4s8[/z](https://www.0crat.com/u/g4s8) (the architect) right in this ticket; if you decide that this PR should not be accepted ever, also inform the architect; this [blog post](http://www.yegor256.com/2015/02/09/serious-code-reviewer.html) will help you understand what is expected from a code reviewer; there will be a monetary reward for this job

**Comment by 0crat** [comment]
Manual assignment of issues is discouraged, see [§19](http://www.zerocracy.com/policy.html#19): -5 point(s) just awarded to @g4s8[/z](https://www.0crat.com/u/g4s8)

File: src/test/java/com/artpie/composer/http/PhpComposerTest.java [hunk 1]
```diff
@@ -77,4 +77,24 @@ void shouldFailPutPackageMetadata() {
...
     }
+
+    @Test
+    void shouldFailGetRootFromNotBasePath() {
+        final Response response = this.php.response(
+            "GET /not-base",
+            Collections.emptyList(),
+            Flowable.empty()
+        );
+        MatcherAssert.assertThat(response, new RsHasStatus(HttpURLConnection.HTTP_NOT_FOUND));
```
Comment (paulodamaso) [inline src/test/java/com/artpie/composer/http/PhpComposerTest.java]: @olegmoz Please, put a failure message and do not use static matchers here

File: src/test/java/com/artpie/composer/http/PhpComposerTest.java [hunk 2]
```diff
@@ -77,4 +77,24 @@ void shouldFailPutPackageMetadata() {
...
+    }
+
+    @Test
+    void shouldFailGetRoot() {
+        final Response response = this.php.response(
+            "GET /base",
+            Collections.emptyList(),
+            Flowable.empty()
+        );
+        MatcherAssert.assertThat(response, new RsHasStatus(HttpURLConnection.HTTP_BAD_METHOD));
```
Comment (paulodamaso) [inline src/test/java/com/artpie/composer/http/PhpComposerTest.java]: @olegmoz Here also

File: src/main/java/com/artpie/composer/http/PhpComposer.java [hunk 3]
```diff
@@ -63,9 +63,13 @@ public Response response(
         final RequestLineFrom request = new RequestLineFrom(line);
         final String path = request.uri().getPath();
         if (path.startsWith(this.base)) {
-            final Resource resource = new PackageMetadata();
-            if (request.method().equals("GET")) {
+            final String relative = path.substring(this.base.length());
+            final Resource resource = PhpComposer.resource(relative);
```
Comment (paulodamaso) [inline src/main/java/com/artpie/composer/http/PhpComposer.java]: @olegmoz Please inline `relative`

**Comment by olegmoz** [comment]
@paulodamaso made changes by your comments, please review

**Comment by olegmoz** [comment]
@paulodamaso there is no implementation for PUT method yet, so writing tests does not make sense.

**Comment by paulodamaso** [comment]
@olegmoz On the contrary: we use test cases to define what behavior is expected from the program, without bothering with the actual implementation. In fact we should always write our tests BEFORE the actual code: having our test cases written beforehand will allow us to assure that our production code is working as intended.

So, the next logical advancement for this thread of issues (adding packages to composer.json, #22) would be writing the tests for the operation (since you've already solved the architectural requirements in this one - #32), and then the PUT operation implementation itself.

That's why we need the puzzle: without it we could not chain the issues properly so they are solved one after another

**Comment by olegmoz** [comment]
@paulodamaso what you are describing is TDD approach when tests are written first and implementation follows. I like this approach, but it is not how it is done in this project. CI requires all tests to pass, so we cannot have any failing tests. So I cannot write proper tests first and writing test that checks method to be failing is just a waste of time.

**Comment by paulodamaso** [comment]
@olegmoz Yes, we should follow TDD approach whenever possible. The resulting code is better and safer.
But our problem here is not testing, is that we had the issue #22 partially solved and we don't have any extension points (a.k.a. todos, or puzzles) to sign the architect and the other developers that the changes asked in #22 were not completely implemented yet. That's why I am asking for the puzzles. We still have two pieces of scope to be done before we can consider the feature asked in #22 as fully implemented: add the PUT operation and add the tests for this operation. And we need to signal it in the code. So that's why I am asking you to put the puzzles. So we can solve these problem in the future, without forgetting about them.
On a side note about TDD, forcing that all the tests pass in the CI is a practice that is not against TDD. It enforces it, actually. You are right, we shouldn't test the behavior of failure in a unimplemented method; we must, instead, write the test for the correct method behavior and do not execute it using the  `@Disabled`  annotation (or any similiar behavior, depending the language you are using). In other words, we implement the test which assures the correct behavior and ignore it in our test run. Then, in the next issue, we implement the method with the correct behavior and activate the test (removing the annotation). This allows us to build the code starting with the tests and without degrading code coverage at all.
So, I'll maintain my request of creating puzzles for the next steps that will need to be done to fully implement #22. Please do it so we can close this issue.

**Comment by olegmoz** [comment]
@paulodamaso added disabled test for PUT method, please review

**Comment by olegmoz** [comment]
@paulodamaso added comment to test, but not going to create puzzle. Changes will be made within same issue, no point in creating new one. We've been over it a lot of times already.

**Comment by paulodamaso** [comment]
@g4s8 All changes related to the code had been addressed, but performer refuses to leave a puzzle to continue this issue implementation. Please we need definition on how intensely we should follow PDD in artipie projects, so these kind of useless discussions stop happening. I just need guidelines  to follow, if we chose to use a flexible form of PDD without using the puzzle formats or whatever i'm ok with this, I just need to know how rigid I have to be in my reviews.

**Comment by g4s8** [comment]
@paulodamaso pdd can be used to close the job to receive payment and continue working on unimplemented parts later, but @olegmoz is a full-timer and doesn't receive payments for jobs. So it doesn't matter how to work with tickets in that case: either use puzzles or submit multiple pull requests to fix one issue.

**Comment by paulodamaso** [comment]
@g4s8 It's ok for me, maybe we should need some directions about it in the README.md then; it is good to merge, then

**Pull request merged**